Duotone: Limit SVG filter output to used filters

jeryj opened the pull request:
Rework of https://github.com/WordPress/gutenberg/pull/48374
Fixes https://github.com/WordPress/gutenberg/issues/38299.
Original PR that was reverted: https://github.com/WordPress/gutenberg/pull/48995

## What?
This PR only adds the needed duotone code to the frontend output if that duotone definition is being used.

## Why?
There are currently eight globally defined duotone styles, and all of their CSS and SVG definitions are output on the frontend of a WordPress website, even if no images are using the duotone filters. The base raw HTML from TT3 on the Sherbert theme variation results in a 76.4kb request. Without any SVG or duotone-related code, the same page HTML is 67.4kb, which means there is up to 9kb of unnecessary duotone-related HTML on every request.

![raw html view of unnecessary SVG code in html](https://user-images.githubusercontent.com/967608/225126237-249a0f42-ff88-4d02-99be-e4b675cdcb2c.png)


## How?
There's quite a bit of changes required to make this happen. We've gone with a route that adds a new `WP_Duotone` class that handles what filters need to be added to the page based on what duotone filters, if any, are in use. Here's a general overview of the high-level changes being made, and why:

### 1. Remove all duotone output from global styles
These are handled by removing the actions related to SVG filters:

```
remove_action( 'wp_body_open', 'wp_global_styles_render_svg_filters' );
remove_action( 'in_admin_header', 'wp_global_styles_render_svg_filters' );
```

### 2. Register all duotone filters from global styles and store in the WP_Duotone class
 
`WP_Duotone::$global_styles_presets` is an array that scrapes all of the global styles duotone filters. These could be defined from theme.json globally, per theme, or in custom global styles. We need to do this to have quick access to the duotone filter definitions if we come across a block that needs to access these styles.

### 3. Register all block-level duotone filters from global styles and store in the WP_Duotone class

`WP_Duotone::$global_styles_block_names` is an array that scrapes all of the blocks that use a duotone filter from global styles, and stores the slug of the duotone filter it uses, such as `default-filter`, `blue-orange`, etc. We need to do this in order to decide if a block being rendered has a duotone filter we need to apply to it.

### 4. Process all blocks to determine the correct duotone output needed

This is the same as before, but we're only outputting the duotone code if needed. On the 'render_block' filter, we process every block and determine:
- Is there a duotone filter we need to apply on this block
- If not, return early
- If we do, decide which duotone filter we need and how to build it
- Add the duotone info we need to `WP_Duotone::$output`

### 5. Output the CSS
On the `'wp_enqueue_scripts' action, output all the necessary CSS for duotone filters collected from `WP_Duotone::$output`.

### 6. Output the SVG
On the `'wp_footer' action, output all necessary SVG for duotone filters from `WP_Duotone::$output`.


<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

## Testing Instructions
These are the different types of Duotone that need testing:
1. Set the duotone in Global Styles to all blocks of a certain type.
2. Set the duotone on a per blocks basis.

For each of these it possible to set the duotone to a `preset` value, a custom value, or `unset`.
For presets, it is possible to set them using two different formats: `var:preset|duotone|blue-orange` or `var(--wp--preset--duotone--blue-orange)`.

It is also important to test that _only_ the SVG filters and CSS that is needed on the page is output - rather than outputting all the presets even if they aren't used.

To test an individual block:
1. Add an image to a post
2. Click on the "filter" option in the block toolbar
3. Select a Duotone option
4. Save the post
5. Open the post in the frontend and check that the correct Duotone filter has been applied

To test Global Styles
1. Open the Global Styles interface in the Site Editor
2. Open "Blocks"
3. Find the image block
4. Click on "filter"
5. Select a Duotone filter
6. Save the Global Styles settings
7. Open the site in the frontend and check that all image blocks have the filter applied (unless they specify their own filters)

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

jeryj committed:
Adds WP_Duotone class to set-up preset and theme-defined blocks using duotone presets

jeryj committed:
Fix incorrect filter name

jeryj committed:
Check if duotone filter is applied on a block in theme.json

This doesn't do anything at the moment. On block render, it checks if the block has duotone filters set via theme.json/global styles/custom styles and returns early if not.

jeryj committed:
Rename duotone class static variables and functions

jeryj committed:
Set WP_Duotone:: array and only output those via wp_footer

jeryj committed:
Only output used duotone presets

Co-authored-by: scruffian <ben@escruffian.com>

jeryj committed:
Cleanup, removal, and only output used SVG filters and CSS

jeryj committed:
Support var:preset|duotone|default-filter syntax for block name styles

jeryj committed:
Refactor to handle both duotone pipe and css var style syntax for duotone preset

jeryj committed:
Renamed output_presets to output, as it also includes custom dutone output. Name could still be better.

jeryj committed:
Fix theme json vs block styles duotone order of application

Refactored out duotone specific code from theme json gutenberg globla styles. Switched theme_json_register_declarations from action to filter.

jeryj committed:
Renamed filter to be more accurate, but should refactor out

Doesn't seem like a legitimately useful filter at the moment. Only for us to be able to use global styles in our duotone-specific output.

jeryj committed:
Simplify getting global styles presets

We were already getting the presets via the global settings, so the filter was unnecessary.

jeryj committed:
Undo theme json class duotone refactoring and remove filter

The filter isn't necesary for us to access what we need. I've removed the refactor as it would be a larger change that can be done in a different PR if needed.

jeryj committed:
Remove unused code

jeryj committed:
Rename save_ to set_ class method names

jeryj committed:
Fix duotone test to use real global preset value

jeryj committed:
Tests for WP_Duotone::gutenberg_get_slug_from_attr()

jeryj committed:
Rework gutenberg_get_slug_from_attr to use regex for more accurate matches

jeryj committed:
Tests for WP_Duotone::is_preset

jeryj committed:
Rework is_preset to also check if the preset exists

jeryj committed:
Remove array_key_exists checks for presets within the block filter

jeryj committed:
Return early if no block content

For some reason, the render_block filter runs even on blocks that are not included on the page. So, if you have a duotone filter defined in your theme.json, such as for the core/post-featured-image block, it will run the block filter even if no featured image is set. As a result, the duotonen defined in theme.json per block would get output even if that block wasn't in use. The only reliable check I could find to determine if the block was in use or not was to check the block content. By bailing early if no block content, we avoid adding unused SVG to the output.

jeryj committed:
Make linter happy

jeryj committed:
Remove unnecessary methods from theme json class

There was some duotone specific code within the theme json class that was only used to build a css declaration for duotone css output. It is now handled manually for speed and simplicity.

jeryj committed:
Renamed get_css_var to get_css_custom_property name and added get_css_var to be more accurately named

Also used these functions throughout duotone.php to refactor for simplicity/consistency. Renamed some declaration variables as well to be more accurately named.

jeryj committed:
Renamed get_css_declaration to get_css_custom_property_declaration

Added a few comments for accuracy

jeryj committed:
Rename WP_Duotone to WP_Duotone_Gutenberg to match gutenberg coding styles

ajlende commented on lib/block-supports/duotone.php:
```diff
@@ -550,6 +541,81 @@ static function () use ( $filter_svg, $selector ) {
+	11
+);
+
+/**
+ * Safari renders elements incorrectly on first paint when the SVG filter comes after the content that it is filtering,
+ * so we force a repaint with a WebKit hack which solves the issue.
+ *
+ * @param string $selector The selector to apply the hack for.
+ */
+function duotone_safari_rerender_hack( $selector ) {
+	/*
+	 * Simply accessing el.offsetHeight flushes layout and style
+	 * changes in WebKit without having to wait for setTimeout.
+	 */
+	printf(
+		'<script>( function() { var el = document.querySelector( %s ); var display = el.style.display; el.style.display = "none"; el.offsetHeight; el.style.display = display; } )();</script>',
+		wp_json_encode( $selector )
+	);
+}
```
This needs to move into `WP_Duotone_Gutenberg`.

jeryj replied:
As a follow-up, I think it'd be good to run this all at once by passing everything from WP_Duotone_Gutenberg::$output and having one <script> tag that handles rerendering all of them.

ajlende commented on lib/block-supports/duotone.php:
```diff
@@ -550,6 +541,81 @@ static function () use ( $filter_svg, $selector ) {
+ */
+add_action(
+	'wp_enqueue_scripts',
+	static function() {
+
+		if ( empty( WP_Duotone_Gutenberg::$output ) ) {
+			return;
+		}
+
+		$duotone_css_vars = '';
+
+		foreach ( WP_Duotone_Gutenberg::$output as $filter_data ) {
+			if ( ! array_key_exists( $filter_data['slug'], WP_Duotone_Gutenberg::$global_styles_presets ) ) {
+				continue;
+			}
+
+			$duotone_css_vars .= WP_Duotone_Gutenberg::get_css_custom_property_declaration( $filter_data );
+		}
+
+		if ( ! empty( $duotone_css_vars ) ) {
+			wp_add_inline_style( 'global-styles', 'body{' . $duotone_css_vars . '}' );
+		}
+	},
```
Maybe the functions for this and the `wp_footer` above could be moved into `WP_Duotone_Gutenberg` too.

If we do, we may be able to make `WP_Duotone_Gutenberg::$output` private to the class as well to minimize the public API.

ajlende replied:
If we change this, we'll also have to move `gutenberg_render_duotone_support` into the class as well.

jeryj replied:
Yeah, it's completely reliant on WP_Duotone_Gutenberg::$output, so let's let the class handle it.

jeryj replied:
> If we change this, we'll also have to move gutenberg_render_duotone_support into the class as well.

I moved the `wp_footer` and `wp_enqueue_scripts` actions into the WP_Duotone_Gutenberg class and it appears to be working fine. Why do we need to move `gutenberg_render_duotone_support` as well? Is it to be able to make it private?

jeryj replied:
Moved it anyways, btw.

ajlende replied:
> Why do we need to move `gutenberg_render_duotone_support` as well? Is it to be able to make it private?

It was using `WP_Duotone_Gutenberg::$output`, so yes, it would have to be part of the class if `$output` is private.

ajlende commented on lib/compat/wordpress-6.2/script-loader.php:
Should these changes be moved to a 6.3 compat file now instead?

jeryj committed:
Move safari rerender hack into WP_Duotone_Gutenberg

jeryj committed:
Move footer action to output SVG into WP_Duotone_Gutenberg.

The footer action that output all the footer resources was fully dependent on the WP_Duotone_Gutenberg class, so we should let it handle this action.

jeryj committed:
Move wp_enqueue_scripts action into WP_Duotone_Gutenberg class.

This was fully reliant on the WP_Duotone_Gutenberg::$output, so let the class handle outputting these styles.

jeryj committed:
Moved gutenberg_render_duotone_support into WP_Duotone_Gutenberg class.

This was very reliant on that class and did the set-up to make it work, so moving it all in to encapsulate it better.

jeryj committed:
Move scrip-loader modifications to 6.3 release

ajlende commented:
@oandregal We're moving the global styles generation code for duotone [into the block supports](https://github.com/WordPress/gutenberg/pull/49103/files#diff-c6cf688aa7b50316afa7fc57327b6ac81c1ac831226447c53ffc88c32fc23750R318-R323) so we can check which filters are being used in `render_block` and only output those to the page.

Could you do a confidence check on if this is a reasonable way for global styles to work and if we're missing anything important that global styles has that we haven't replicated in the duotone block supports? Thanks!

jeryj committed:
Set WP_Duotone_Gutenberg static variables private

jeryj committed:
Linting and comments

jeryj committed:
Switched to protected static from private static

ajlende commented on lib/block-supports/duotone.php:
```diff
@@ -421,143 +421,10 @@ function gutenberg_register_duotone_support( $block_type ) {
- * @param  array  $block         Block object.
- * @return string                Filtered block content.
- */
-function gutenberg_render_duotone_support( $block_content, $block ) {
```
This function probably still needs to exist because it was public, but it can just call the class-based one.

ajlende replied:
There's a [`_deprecated_function`](https://developer.wordpress.org/reference/functions/_deprecated_function/) function that can be called to mark this one as deprecated in favor of the class-based one.

ajlende commented on lib/class-wp-duotone-gutenberg.php:
```diff
@@ -0,0 +1,375 @@
+	 * @since 6.3.0
+	 * @var array
+	 */
+	protected static $global_styles_presets = array();
```
I think `private` is fine. In global styles classes, we made things `protected` so we could extend the class rather than copying things over, but that ended up being messier so we've switched back now.

aaronrobertshaw replied:
For those playing along at home, https://github.com/WordPress/gutenberg/pull/46579 is the PR switching the `WP_Theme_JSON` override class to being bundled. Following the same approach for this new duotone class sounds good 👍

ajlende commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -144,7 +144,7 @@ class WP_Theme_JSON_Gutenberg {
 			'prevent_override'  => array( 'color', 'defaultDuotone' ),
 			'use_default_names' => false,
-			'value_func'        => 'gutenberg_get_duotone_filter_property',
+			'value_func'        => null, // Don't output CSS Custom Properties for duotone.
 			'css_vars'          => '--wp--preset--duotone--$slug',
```
```suggestion
			'value_func'        => null, // CSS Custom Properties for duotone are handled by block supports in class-wp-duotone-gutenberg.php.
			'css_vars'          => null,
```

@oandregal correct me if I'm wrong about how this should be handled. Should the `properties` array also be emptied?

ajlende replied:
This will probably result in some unused code paths in global styles, but as this is already a pretty large change to review the plan is to investigate that in a follow-up.

aaronrobertshaw replied:
I assume we'll add a new issue for the follow-up to the [Duotone filters](https://github.com/WordPress/gutenberg/issues/47304) tracking issue?

ajlende replied:
Yeah, I'll be updating it after this is merged with any follow-ups that we may need.

ajlende commented on lib/class-wp-duotone-gutenberg.php:
```diff
@@ -0,0 +1,375 @@
+	 * @return bool True if the duotone preset present and valid.
+	 */
+	static function is_preset( $duotone_attr ) {
+		$slug = WP_Duotone_Gutenberg::gutenberg_get_slug_from_attr( $duotone_attr );
```
We should probably prefer `self` instead of the class name here and throughout the rest of the class.

```suggestion
		$slug = self::gutenberg_get_slug_from_attr( $duotone_attr );
```

ajlende commented on lib/class-wp-duotone-gutenberg.php:
```diff
@@ -0,0 +1,375 @@
+	 * @param string $slug The slug of the duotone preset.
+	 * @return string The CSS variable name.
+	 */
+	static function get_css_custom_property_name( $slug ) {
```
Functions like this that are only used within the class should be `private`.

```suggestion
	private static function get_css_custom_property_name( $slug ) {
```

ajlende commented on lib/class-wp-duotone-gutenberg.php:
```diff
@@ -0,0 +1,375 @@
+				array(
+					'selector'     => $selector,
+					'declarations' => array(
+						// !important is needed because these styles
+						// render before global styles,
+						// and they should be overriding the duotone
+						// filters set by global styles.
+						'filter' => $declaration_value . ' !important',
```
This comment is now out of date since we aren't generating duotone styles in global styles anymore.

aaronrobertshaw replied:
If we aren't generating duotone styles in global styles anymore, can we also do away with the `! important`?

jeryj replied:
It seems like we should be able to, but when I remove the `!important`, the CSS rule isn't output on the page at all. I'm not sure why though. Something is stripping it out or not allowing it be added. Looking into it.

ajlende replied:
> Something is stripping it out or not allowing it be added.

That's related to the kses filter being especially specific for this feature in https://github.com/WordPress/gutenberg/pull/48281/files/08d51b9dd8edc80f32bc035f7f8b6e5931508b70..8dddbd21f2aa9a5224e5214ebe3e9cfee82d386b#diff-9cac08004c469d02cbe46472d33dfb2349d2694d223be7bad4a59d4204a1702d until the more general `filter` property is added to the list of kses properties that support `url()` in https://github.com/WordPress/wordpress-develop/pull/4108.

ajlende commented:
@aaronrobertshaw and @andrewserong This may be of interest to you and maybe others on your team as well since it is doing some similar things to the style engine and may be able to leverage the style engine more.

ajlende commented on lib/block-supports/duotone.php:
```diff
@@ -421,143 +421,10 @@ function gutenberg_register_duotone_support( $block_type ) {
- * @param  array  $block         Block object.
- * @return string                Filtered block content.
- */
-function gutenberg_render_duotone_support( $block_content, $block ) {
```
For those reviewing, this function had modifications and was moved to the `WP_Duotone_Gutenberg` class. It may be a little easier to see the modifications in [this diff with an earlier commit](https://github.com/WordPress/gutenberg/compare/trunk...737b9cb7ee06ebe44a09748abaa8a1d6062916c7?w=1).

ajlende commented on lib/class-wp-duotone-gutenberg.php:
```diff
@@ -0,0 +1,375 @@
+	}
+}
+
+add_action( 'wp_loaded', array( 'WP_Duotone_Gutenberg', 'set_global_styles_presets' ), 10 );
+add_action( 'wp_loaded', array( 'WP_Duotone_Gutenberg', 'set_global_style_block_names' ), 10 );
+// Remove WordPress core filter to avoid rendering duplicate support elements.
+remove_filter( 'render_block', 'wp_render_duotone_support', 10, 2 );
+add_filter( 'render_block', array( 'WP_Duotone_Gutenberg', 'render_duotone_support' ), 10, 2 );
+add_action( 'wp_enqueue_scripts', array( 'WP_Duotone_Gutenberg', 'output_global_styles' ), 11 );
+add_action( 'wp_footer', array( 'WP_Duotone_Gutenberg', 'output_footer_assets' ), 10 );
```
I think these actions/filters probably still belong in block-supports/duotone.php

aaronrobertshaw replied:
Agreed 👍

jeryj replied:
Happy to move them over. For my own learning, can you explain why? I'm guessing it's because `block-supports/duotone.php` should be the thing that _uses_ the `WP_Duotone_Gutenberg` class to provide the duotone functionality.

aaronrobertshaw replied:
That about sums it up for me. It's also then consistent with all the other block supports and where they manage their filters.

ajlende replied:
> For my own learning, can you explain why? 

From my perspective, every action is a side-effect which means just importing the duotone class with these would cause the actions to be added to the page. It's expected that the block-supports files will have these side-effects, but it isn't expected to import a class and have side-effects happen.

ajlende commented on phpunit/block-supports/duotone-test.php:
```diff
@@ -34,7 +34,44 @@ public function test_gutenberg_render_duotone_support_custom() {
+		return array(
+			'pipe-slug'                       => array( 'var:preset|duotone|blue-orange', 'blue-orange' ),
+			'css-var'                         => array( 'var(--wp--preset--duotone--blue-orange)', 'blue-orange' ),
+			'css-var-weird-chars'             => array( 'var(--wp--preset--duotone--.)', '.' ),
```
See https://github.com/WordPress/gutenberg/pull/48995#discussion_r1135899376

```suggestion
			'css-var-invalid-slug-chars'      => array( 'var(--wp--preset--duotone--.)', '.' ),
```

And I'd do the same rename below.

aaronrobertshaw commented on lib/class-wp-duotone-gutenberg.php:
```diff
@@ -0,0 +1,375 @@
+	protected static $global_styles_block_names = array();
+
+	/**
+	 * An array of Duotone SVG and CSS ouput needed for the frontend duotone rendering based on what is
+	 * being ouptput on the page. Organized by a slug of the preset/color group and the information needed
```
```suggestion
	 * An array of Duotone SVG and CSS output needed for the frontend duotone rendering based on what is
	 * being output on the page. Organized by a slug of the preset/color group and the information needed
```

Couple of small typos.

aaronrobertshaw commented on lib/class-wp-duotone-gutenberg.php:
```diff
@@ -0,0 +1,375 @@
+	 */
+	protected static $output = array();
+
+
```
Nit: Excess blank line.

aaronrobertshaw commented on lib/class-wp-duotone-gutenberg.php:
```diff
@@ -0,0 +1,375 @@
+			WP_Duotone_Gutenberg::$output[ $slug ]['selector'] = $selector;
+		}
+
+		// Calling gutenberg_style_engine_get_stylesheet_from_css_rules ensures that
+		// the styles are rendered in an inline for block supports because we're
+		// using the `context` option to instruct it so.
```
Typo: Could we improve the "rendered in an inline" wording?

jeryj replied:
@ajlende Do you think this comment is still relevant? I'm not sure why we need to ensure it's in this inline block support style section. Or is it that it only needs to be output after the global styles?

ajlende replied:
> Do you think this comment is still relevant?

The way `gutenberg_style_engine_get_stylesheet_from_css_rules` is used hasn't been changed by this PR. It's still handling all of the class generation for duotone. So I assume it's still relevant.

> I'm not sure why we need to ensure it's in this inline block support style section.

I'm not sure why either. @andrewserong gave it the check when it was added, so he may be able to explain?

> Or is it that it only needs to be output after the global styles?

It's not really a matter of before or after in this particular case. It might help if I try to explain how this part of the code is related to global styles.

Most global styles have classes that can be applied to the block wrapper because the CSS properties they set are inherited properties. Global styles generates these based on the `'classes'` key in `PRESETS_METADATA` in [class-wp-theme-json-gutenberg.php](https://github.com/WordPress/gutenberg/blob/55e16d3c8a0e3b2ee8fd7f87a12cae8a2b1886d0/lib/class-wp-theme-json-gutenberg.php#L127). On every page, all of these class names are generated whether or not they are used. The mental model I have is that it works like all of the old-school CSS frameworks (Blueprint, Logic, YUI, etc.) that are just a single sheet of CSS that includes everything, but for entire blocks instead of just individual HTML elements.

Duotone is different because `filter` is a non-inherited CSS property which makes the CSS that needs to be generated dependent on the structure of each specific block. In hindsight I could have generated all the combinations of presets and registered block selectors to fit that paradigm. But that never even crossed my mind and it would have been even more wasteful than what what we're trying to fix now.

So what we ended up with was global styles classes for duotone being generated on block render. And that's why the function of this particular bit of code is still fulfilling the same role as it was prior to this PR.

andrewserong replied:
Thanks for the ping!

For a bit of context surrounding the `context` prop 😄 — prior to the style engine being introduced, block supports would output a style tag for each instance of a block being rendered, which resulted in numerous redundant style tags being output to the site frontend.

The style engine's `context` prop allows styles to be grouped together into a single stylesheet. In this case, I think it's fine to leave it with the `block-supports` context, as the style output for block supports is already handled, and unless there's a need to have a separate stylesheet for Duotone, I think it makes sense to combine into the one stylesheet. Separately, there are efforts being explored to roll out the style engine for global styles (e.g. in https://github.com/WordPress/gutenberg/pull/48955), but it should be pretty simple to change the `context` value here in a follow-up further down the track if we need to.

> Typo: Could we improve the "rendered in an inline" wording?

Perhaps it could be worded more like:

"Pass styles to the block-supports stylesheet via the style engine. This ensures that Duotone styles are included in a single stylesheet, avoiding multiple style tags or multiple stylesheets being output to the site frontend."

aaronrobertshaw commented on lib/class-wp-duotone-gutenberg.php:
```diff
@@ -0,0 +1,375 @@
+			)
+		);
+
+		// Like the layout hook, this assumes the hook only applies to blocks with a single wrapper.
+		return preg_replace(
+			'/' . preg_quote( 'class="', '/' ) . '/',
+			'class="' . $filter_id . ' ',
+			$block_content,
+			1
+		);
```
Would leveraging the `WP_HTML_Tag_Processor` potentially offer a more robust solution here?

jeryj replied:
Are you OK with this conversation in a follow-up PR? This behavior isn't changed, and I'm worried about how large/many changes this PR introduces already.

ajlende replied:
Yeah, this can be a follow-up that I'll add to the [Duotone filters](https://github.com/WordPress/gutenberg/issues/47304) tracking issue once this is merged.

aaronrobertshaw replied:
No objections here 👍

aaronrobertshaw commented on lib/compat/wordpress-6.3/script-loader.php:
```diff
@@ -0,0 +1,15 @@
+ */
+
+/**
+ * Removes duotone SVG filter presets from all being output.
```
```suggestion
 * Prevent unused duotone SVG filter presets from being output.
```

It could be the hour of the day but I had to read this sentence twice before understanding its intended implication. Maybe we could reword it?

jeryj replied:
I can see how that line is confusing. These remove the action that outputs all of the SVGs, so that the WP_Duotone_Gutenberg class can handle outputting only the used SVGs. I'll update the comment to explain this more clearly and maybe that'll help give more context to make it clearer. Happy to keep tweaking this though!

aaronrobertshaw reviewed (commented):
> This may be of interest to you and maybe others on your team as well since it is doing some similar things to the style engine and may be able to leverage the style engine more.

Thanks for the ping @ajlende 👍 

@andrewserong is probably best placed to see opportunities to leverage, or improve, the style engine more.

I've given this a quick look over and only spotted minor issues in addition to the feedback already left by others. I hope to get some extra time to test this more thoroughly as it evolves.

One last question, would it be beneficial to split the `WP_Duotone` class-specific unit tests out into their own file? I expected to see a test file with a similar name to the class 🤷

jeryj committed:
Minor cleanup and comments

jeryj committed:
Fix comment misspellings

Co-authored-by: Aaron Robertshaw <60436221+aaronrobertshaw@users.noreply.github.com>

jeryj committed:
Change from protected to private variables

Prevents us from unintentionally needing to support the usage of these variables publicly on a rapidly changing api

jeryj committed:
Use self:: instead of WP_Duotone_Gutenberg:: within class

jeryj committed:
Move actions from class into duotone.php

jeryj committed:
Set internal methods to private in duotone class

jeryj committed:
Rename duotone weird-chars test to invalid-slug-chards

jeryj committed:
Add gutenberg_render_duotone_support as deprecated

jeryj committed:
Rename duotone test file to class-wp-duotone-test

jeryj committed:
Update tests to allow access to private static methods

jeryj committed:
Rename test suite to WP_Duotone_Gutenberg_Test

jeryj commented:
> One last question, would it be beneficial to split the WP_Duotone class-specific unit tests out into their own file? I expected to see a test file with a similar name to the class 🤷

Done ✅

jeryj committed:
Set wp theme json duotone value_func to null

Co-authored-by: Alex Lende <alex@lende.xyz>

ajlende commented on lib/block-supports/duotone.php:
```diff
@@ -421,133 +421,17 @@ function gutenberg_register_duotone_support( $block_type ) {
-		$block_content,
-		1
-	);
+	_deprecated_function( __METHOD__, '6.3.0', 'WP_Duotone_Gutenberg::render_duotone_support' );
```
```suggestion
	_deprecated_function( __FUNCTION__, '6.3.0', 'WP_Duotone_Gutenberg::render_duotone_support' );
```

jeryj committed:
Change deprecated magic constant to correct name of __FUNCTION__

Co-authored-by: Alex Lende <alex@lende.xyz>

andrewserong commented on lib/class-wp-duotone-gutenberg.php:
```diff
@@ -0,0 +1,366 @@
+	}
+
+	/**
+	 * Render out the duotone stylesheet and SVG.
```
Tiny nit: since this is using the style engine to output styles within an existing stylesheet, should this refer to CSS styles instead of a separate stylesheet?

```suggestion
	 * Render out the duotone CSS styles and SVG.
```

andrewserong commented:
> @andrewserong is probably best placed to see opportunities to leverage, or improve, the style engine more.

Thanks for the ping! At this stage, I think this PR's doing a good job of using the existing style engine function where it's useful, I don't think there's anything I'd change at this point in regards to the current style engine work. Eventually, once the style engine is incorporated into global styles output, there might be an opportunity to refactor the internals of the `output_global_styles` method, but I imagine that would be a pretty minor change. @aristath has kicked off some of the work for the style engine in global styles in a WIP over in https://github.com/WordPress/gutenberg/pull/48955.

jeryj committed:
Update comment about how duotone styles are included in the stylesheet

Co-authored-by: Andrew Serong <andrewserong@users.noreply.github.com>

jeryj committed:
Correct duotone stylehseet to dutuotone CSS styles comment

Co-authored-by: Andrew Serong <14988353+andrewserong@users.noreply.github.com>

jeryj committed:
Lint php

ajlende commented on lib/class-wp-duotone-gutenberg.php:
```diff
@@ -0,0 +1,367 @@
+	 */
+	private static $output = array();
+
+	const CSS_VAR_PREFIX = '--wp--preset--duotone--';
```
One last nit. I don't think this necessarily has to be `private` as just a constant string, but PHP defaults to `public` and it wasn't explicitly labeled as such.

```suggestion
	/**
	 * Prefix used for generating and referencing duotone CSS custom properties.
	 */
	private const CSS_VAR_PREFIX = '--wp--preset--duotone--';
```

jeryj committed:
Set WP_Duotone_Gutenberg CSS_VAR_PREFIX to private

Co-authored-by: Alex Lende <alex@lende.xyz>

ajlende reviewed (approved):
I keep getting tripped up by [broken preset filters in the editor #48586](https://github.com/WordPress/gutenberg/issues/48586) when going through my usual bout of tests, but this is only changing the frontend part of the generation so that isn't of concern here.

I'm happy to get this merged and open up some follow-ups 👍

Commit:
Remove private specifier to CSS_VAR_PREFIX

It isn't supported in PHP 7.0

jeryj committed:
Added tests teardown to prevent css rules from leaking to other phpunit tests

oandregal commented on lib/class-wp-duotone-gutenberg.php:
```diff
@@ -0,0 +1,370 @@
+	 */
+	public static function set_global_style_block_names() {
+		// Get the per block settings from the theme.json.
+		$tree        = WP_Theme_JSON_Resolver::get_merged_data();
```
@jeryj @ajlende @aaronrobertshaw @andrewserong We need to use the class bundled by Gutenberg instead of the core class. Fix for this at https://github.com/WordPress/gutenberg/pull/49199

ajlende replied:
Good catch! Thanks for getting that fix out!

aaronrobertshaw replied:
@oandregal @ajlende @jeryj it appears that the fix in https://github.com/WordPress/gutenberg/pull/49199 surfaced an issue with the duotone action [added in this PR](https://github.com/WordPress/gutenberg/pull/49103/files#diff-fd626ca219f3d805667d480eabb2d890a5a8d2e015b37f03cf8980edf1ae3d58R446), that causes theme.json block node data to be cached too soon preventing the editor-only styles introduced by https://github.com/WordPress/gutenberg/pull/46496.

While debugging on trunk, I found that global styles duotone filters aren't being applied in either the post editor (https://github.com/WordPress/gutenberg/pull/49239) or the frontend. With regards to https://github.com/WordPress/gutenberg/pull/49239, that branch works until you rebase it on trunk where it then fails again to apply the duotone filter.

_Update: Found that the `image/block.json` file needs the `filters.duotone` in the selectors now for the global styles and probably some other tweaks to allow moving away from `support.color.__experimentalDutone`._

I wasn't able to find another way to detect we're in the editor for editor-only selectors than via `get_current_screen` and the duotone action here appears to be getting the block node data before that function exists.

I'll create an issue for this but also wanted to flag it within the context of this PR.

_Edit: New issue - https://github.com/WordPress/gutenberg/issues/49324_
